core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,403 @@
+            exploration_versioned = exp_services.get_exploration_by_id(
+                exp_id, version)
+
+            if version == 1:
+                # Create default state stats mapping for the first version.
+                for state_name in exploration_versioned.states:
+                    state_stats_mapping[state_name] = (
+                        stats_domain.StateStats.create_default())
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                        state_stats_mapping[change_dict['state_name']] = (
+                            stats_domain.StateStats.create_default())
+                    elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                        state_stats_mapping.pop(change_dict['state_name'])
+                    elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                        state_stats_mapping[change_dict['new_state_name']] = (
+                            state_stats_mapping.pop(
+                                change_dict['old_state_name']))
+
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in exploration_versioned.states
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for value in values_state_hit_versioned:
+                state_name = value['state_name']
+                state_stats_mapping[state_name].total_hit_count_v1 += 1
+                if value['session_id'] not in state_session_ids[state_name]:
+                    state_stats_mapping[state_name].first_hit_count_v1 += 1
+                    state_session_ids[state_name].append(value['session_id'])
+                    session_ids_set.add(value['session_id'])
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                values_state_hit_sessioned = [
+                    val for val in values_state_hit_versioned if val[
+                        'session_id'] == session_id]
+                # All states in the above list except the last element (only if
+                # complete exploration event doesn't exist for the session ID)
+                # are counted as completed because they are ordered by their
+                # timestamp.
+                if session_id not in completed_session_ids:
+                    for value in values_state_hit_sessioned[:-1]:
```

Comment:
Pull the values_state_hit_sessioned[:-1] out of the if/else. Have the if/else handle only the values_state_hit_sessioned[-1] case.